src/language_features/document_links.rs

```diff
@@ -0,0 +1,213 @@
+                server_id,
+                vec![DocumentLinkParams {
+                    partial_result_params: Default::default(),
+                    text_document: TextDocumentIdentifier {
+                        uri: Url::from_file_path(&meta.buffile).unwrap(),
+                    },
+                    work_done_progress_params: Default::default(),
+                }],
+            )
+        })
+        .collect();
+    ctx.call::<DocumentLinkRequest, _>(
+        meta,
+        RequestParams::Each(req_params),
+        move |ctx, meta, results| {
+            document_links_response(meta, results, ctx);
+        },
+    );
+}
+
+fn document_links_response(
+    meta: EditorMeta,
+    results: Vec<(ServerId, Option<Vec<DocumentLink>>)>,
+    ctx: &mut Context,
+) {
+    let mut links = results
+        .into_iter()
+        .flat_map(|(server_id, v)| {
+            let v: Vec<_> = v
+                .unwrap_or_default()
+                .into_iter()
+                .map(|v| (server_id, v))
+                .collect();
+            v
+        })
+        .collect::<Vec<_>>();
+    links.sort_by_key(|(_, link)| link.range.start);
+
+    let buffile = &meta.buffile;
+    let document = match ctx.documents.get(buffile) {
+        Some(document) => document,
+        None => {
+            ctx.document_links.remove(buffile);
+            return;
+        }
+    };
+    let version = document.version;
+    let range_specs = links
+        .iter()
+        .map(|(server_id, link)| {
+            let server = ctx.server(*server_id);
+            format!(
+                "{}|LspDocumentLink",
+                lsp_range_to_kakoune(&link.range, &document.text, server.offset_encoding)
+            )
+        })
+        .join(" ");
+
+    ctx.document_links.insert(buffile.clone(), links);
+
+    let command = format!("set-option buffer lsp_document_links {version} {range_specs}");
+    let command = format!(
+        "evaluate-commands -buffer {} %§{}§",
+        editor_quote(buffile),
+        command.replace('§', "§§")
+    );
+    ctx.exec(EditorMeta::default(), command);
+}
+
+#[derive(Clone, Debug)]
+pub struct DocumentLinkOptions {
+    pub selection_desc: String,
+}
+
+pub fn resolve_and_follow_document_link(
+    meta: EditorMeta,
+    params: DocumentLinkOptions,
+    ctx: &mut Context,
+) {
+    let (range, _cursor) = parse_kakoune_range(&params.selection_desc);
+    let document = match ctx.documents.get(&meta.buffile) {
+        Some(document) => document,
+        None => return,
+    };
+
+    if let Some((server_id, link)) = ctx
+        .document_links
+        .get(&meta.buffile)
+        .and_then(|links| {
+            links.iter().find(|(server_id, link)| {
+                let ServerSettings {
+                    offset_encoding, ..
+                } = ctx.server(*server_id);
+                let range = kakoune_range_to_lsp(&range, &document.text, *offset_encoding);
+                ranges_overlap(link.range, range)
```

Comment:
should probably deduplicate this `filter()` call (maybe extract a function if we stick with two separate code paths here)